pkg/util/validation/limits.go

```diff
@@ -831,8 +831,19 @@ func (o *Overrides) RulerTenantShardSize(userID string) int {
 }
 
 // RulerMaxRulesPerRuleGroup returns the maximum number of rules per rule group for a given user.
-func (o *Overrides) RulerMaxRulesPerRuleGroup(userID string) int {
-	return o.getOverridesForUser(userID).RulerMaxRulesPerRuleGroup
+// This limit is special. Limits are returned in the following order:
+// 1. Per tenant limit for the given namespace.
+// 2. Default limit for the given namespace.
+// 3. Per tenant limit set by RulerMaxRulesPerRuleGroup
+// 4. Default limit set by RulerMaxRulesPerRuleGroup
+func (o *Overrides) RulerMaxRulesPerRuleGroup(userID, namespace string) int {
+	u := o.getOverridesForUser(userID)
+
+	if namespaceLimit, ok := u.RulerMaxRulesPerRuleGroupByNamespace.data[namespace]; ok {
+		return int(namespaceLimit)
```

Comment:
Why would we define this map as `LimitsMap[int64]` if we're going to cast to an `int` later? Can we just define it as `LimitsMap[int]` from the beginning?

Reply:
Yeah, I agree with you: https://github.com/grafana/mimir/pull/8378/commits/6c9a3bff174498dc8bd35764bbf024d8e4362617#diff-1e0a07c0c1ea87ab5a582d79c587a73a9b8e0680d89c61d5c94be106823a4a4e

Reply:
Done in https://github.com/grafana/mimir/pull/8378/commits/7c6c7370e8c8c4c0355c39f6ea03bcfda93294a5